### PR TITLE
Support numerical types in `if` predicate and `not` expression

### PR DESCRIPTION
**Pull request by klecki**
## Description:
Evaluate scalar numerical types treating them as truthy and falsy values.
Utilize regular promotion from numeric types to boolean. We can do it safely as `not` produces a new value and the predicate in `if` is not related to the output.
`and` and `or` expression stay limited to `bool` inputs to not risk creating scenarios where batches of mixed types can arise.

### Affected modules and functionalities:
Conditional expressions, split/merge.

### Tests:

- [x] Existing tests apply
Existing test was modified to not expect exception when non-boolean type is passed as `if` predicate or into `not`.
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [x] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: dali/pipeline/operator/builtin/conditional/split_merge.h [hunk 1]
```diff
@@ -30,10 +33,16 @@ namespace dali {
  * @param condition_idx Index of the sample that this condition applies to (input for split, output
  * for merge)
  */
-inline int get_group_index(const TensorList<CPUBackend>& condition, int condition_idx,
-                              bool is_logical = true) {
+inline int get_group_index(const TensorList<CPUBackend> &condition, int condition_idx,
```
Comment (stiepan) [inline dali/pipeline/operator/builtin/conditional/split_merge.h:36]: Just a food for thought, could it affect perf if we accepted TensorView here and did the switch for the while batch instead as we do most of the time (when samples are far from scalars of course).

File: docs/examples/general/conditionals.ipynb [hunk 2]
```diff
@@ -153,7 +153,7 @@
     "\n",
     "### Requirements\n",
     "There are several requirements that needs to be met when using the conditionals within DALI pipeline:\n",
-    "1. The `if` condition must be a DALI DataNode and the underlying data must be of type Bool.\n",
+    "1. The `if` condition must be a DALI DataNode and the underlying data must be of type `BOOL` or other numerical type that can be evaluated to True or False in accordance with Python semantics.\n",
```
Comment (mzient) [inline docs/examples/general/conditionals.ipynb:156]: I think that at this point there's no reason to mention bool.
```suggestion
    "1. The `if` condition must be a DALI DataNode and the underlying data must be of a numerical type that can be evaluated to True or False in accordance with Python semantics.\n",
```
This is a nitpick / thing to consider, nothing serious.

**Review by mzient — APPROVED** [review]
+1 for updating stale comments

File: docs/examples/general/conditionals.ipynb (hunk 2 above)
Comment (klecki) [reply docs/examples/general/conditionals.ipynb:156]: CI is super lazy, I will update it when I revisit the docs about the logical expressions `and`, `or` and `not`.

**Pull request merged**